### PR TITLE
feat: PG Agnostic mandate using network_txns_id (Adyen, Authorizedotnet, Stripe)

### PR DESCRIPTION
**Pull request by manoj-juspay**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Implemented Adyen mandate using connector_mandate_id
https://juspay.atlassian.net/browse/HYP-37

### Additional Changes

- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
Setup mandate followed by recurring payment

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -298,14 +298,20 @@ pub enum MandateTxnType {
 #[derive(Default, Eq, PartialEq, Debug, serde::Deserialize, serde::Serialize, Clone)]
 pub struct MandateIds {
     pub mandate_id: String,
-    pub connector_mandate_id: Option<String>,
+    pub mandate_reference_id: Option<MandateReferenceId>,
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:301]: Add comments for the ids

File: crates/router/src/core/mandate.rs [hunk 2]
```diff
@@ -164,13 +165,14 @@ where
...
                     mandate_reference,
+                    network_txn_id,
                 ) {
                     let connector = new_mandate_data.connector.clone();
                     logger::debug!("{:?}", new_mandate_data);
                     resp.request
                         .set_mandate_id(api_models::payments::MandateIds {
                             mandate_id: new_mandate_data.mandate_id.clone(),
-                            connector_mandate_id: new_mandate_data.connector_mandate_id.clone(),
+                            mandate_reference_id: None, //This is not used further in case needed please handle
```
Comment (jarnura) [inline crates/router/src/core/mandate.rs:175]: just pass the value based on match logic

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 3]
```diff
@@ -68,31 +68,86 @@ enum PaymentDetails {
...
+            });
+            let subseuent_auth_info = Some(SubsequentAuthInformation {
+                original_network_trans_id,
+                reason: Reason::Resubmission,
+            });
+            match item.request.payment_method_data {
+                api::PaymentMethodData::Card(ref ccard) => {
+                    let payment_details = PaymentDetails::CreditCard(CreditCardDetails {
+                        card_number: ccard.card_number.clone(),
+                        // expiration_date: format!("{expiry_year}-{expiry_month}").into(),
```
Comment (jarnura) [inline crates/router/src/connector/authorizedotnet/transformers.rs:101]: remove commented code

File: crates/router/src/connector/adyen/transformers.rs [hunk 4]
```diff
@@ -4,7 +4,7 @@ use reqwest::Url;
 use serde::{Deserialize, Serialize};
 
 use crate::{
-    connector::utils::PaymentsAuthorizeRequestData,
+    connector::utils::{missing_field_err, PaymentsAuthorizeRequestData},
```
Comment (jarnura) [inline crates/router/src/connector/adyen/transformers.rs:7]: Remove missing_field_err in imports and use like utils::missing_field_err

File: crates/router/src/connector/adyen/transformers.rs [hunk 5]
```diff
@@ -566,98 +611,139 @@ fn get_country_code(item: &types::PaymentsAuthorizeRouterData) -> Option<api_enu
...
+        Some(api_models::payments::MandateReferenceId::NetworkMandateId(network_mandate_id)) => {
+            match item.request.payment_method_data {
+                api::PaymentMethodData::Card(ref card) => {
+                    let adyen_card = AdyenCard {
+                        payment_type: PaymentType::Scheme,
+                        number: card.card_number.clone(),
+                        expiry_month: card.card_exp_month.clone(),
+                        expiry_year: card.card_exp_year.clone(),
+                        cvc: None,
+                        brand: Some(CardBrand::Visa), //Fixme
```
Comment (jarnura) [inline crates/router/src/connector/adyen/transformers.rs:638]: Explain the problem for FIXME. If there is a issue already linked add the issue here.

Comment (jarnura) [inline crates/router/src/connector/adyen/transformers.rs:638]: And we can't go with Hard coding the CardBrand.

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/api_models/src/payments.rs (hunk 1 above)
Comment (manoj-juspay) [reply crates/api_models/src/payments.rs:305]: sure

File: crates/router/src/core/mandate.rs (hunk 2 above)
Comment (manoj-juspay) [reply crates/router/src/core/mandate.rs:175]: ok

File: crates/router/src/connector/authorizedotnet/transformers.rs (hunk 3 above)
Comment (manoj-juspay) [reply crates/router/src/connector/authorizedotnet/transformers.rs:101]: sure, it was there before maybe for some reference

File: crates/router/src/connector/adyen/transformers.rs [hunk 6]
```diff
@@ -246,6 +266,16 @@ pub struct AdyenCard {
     expiry_month: Secret<String>,
     expiry_year: Secret<String>,
     cvc: Option<Secret<String>>,
+    brand: Option<CardBrand>, //Mandatory for mandate using network_txns_id
+    network_payment_reference: Option<String>, //
```
[the remark this answers is not in the document]
Comment (manoj-juspay) [reply crates/router/src/connector/adyen/transformers.rs:516]: ok

File: crates/router/src/connector/adyen/transformers.rs (hunk 5 above)
Comment (manoj-juspay) [reply crates/router/src/connector/adyen/transformers.rs]: sure bro

Comment (manoj-juspay) [reply crates/router/src/connector/adyen/transformers.rs]: handled

File: crates/router/src/connector/adyen/transformers.rs (hunk 4 above)
Comment (manoj-juspay) [reply crates/router/src/connector/adyen/transformers.rs:9]: done

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/stripe/transformers.rs [hunk 7]
```diff
@@ -1427,12 +1483,14 @@ impl TryFrom<&types::PaymentsCancelRouterData> for CancelRequest {
...
 
-#[derive(Deserialize, Debug, Clone, Eq, PartialEq)]
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
 #[non_exhaustive]
 #[serde(rename_all = "snake_case")]
 pub enum StripePaymentMethodOptions {
     Card {
         mandate_options: Option<StripeMandateOptions>,
+        network_transaction_id: Option<String>,
+        mit_exemption: Option<MITExemption>, // To be used for MIT mandate txns
```
Comment (jarnura) [inline crates/router/src/connector/stripe/transformers.rs:1493]:
```suggestion
        mit_exemption: Option<MitExemption>, // To be used for MIT mandate txns
```

File: crates/router/src/connector/stripe/transformers.rs [hunk 8]
```diff
@@ -1450,9 +1508,19 @@ pub enum StripePaymentMethodOptions {
     #[serde(rename = "bacs_debit")]
     Bacs {},
 }
+
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
+pub struct MITExemption {
```
Comment (jarnura) [inline crates/router/src/connector/stripe/transformers.rs:1513]:
```suggestion
pub struct MitExemption {
```

File: crates/router/src/core/payments/flows/authorize_flow.rs [hunk 9]
```diff
@@ -367,6 +378,168 @@ impl mandate::MandateBehaviour for types::PaymentsAuthorizeData {
...
+    connector: &api::ConnectorData,
+    customer: &Option<storage::Customer>,
+) -> RouterResult<CreateCustomerCheck> {
+    let connector_name = connector.connector_name.to_string();
+    let connector_customer_filter = state
+        .conf
+        .connector_customer
+        .connector_list
+        .contains(&connector.connector_name);
+    if connector_customer_filter {
```
Comment (jarnura) [inline crates/router/src/core/payments/flows/authorize_flow.rs:434]: Avoid deep nesting function, split into small function and write a comments too.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/flows.rs [hunk 10]
```diff
@@ -68,6 +69,20 @@ pub trait Feature<F, T> {
...
+    where
+        F: Clone,
+        Self: Sized,
+        dyn api::Connector: services::ConnectorIntegration<F, T, types::PaymentsResponseData>,
+    {
+        Ok(None)
+    }
 }
 
 macro_rules! default_imp_for_complete_authorize{
```
[the remark this answers is not in the document]
Comment (manoj-juspay) [reply crates/router/src/core/payments/flows.rs:88]: this is for flow we only need it for authorize i have changed it in all the connector and added a macro here

File: crates/router/src/types.rs [hunk 11]
```diff
@@ -285,6 +310,11 @@ pub enum PaymentsResponseData {
     TokenizationResponse {
         token: String,
     },
+
+    ConnectorCustomerResponse {
+        connector_cust_id: String,
```
[the remark this answers is not in the document]
Comment (manoj-juspay) [reply crates/router/src/types.rs:315]: sure

File: crates/router/src/connector/stripe/transformers.rs [hunk 12]
```diff
@@ -994,13 +1037,15 @@ pub struct PaymentIntentResponse {
     pub client_secret: Secret<String>,
     pub created: i32,
     pub customer: Option<String>,
+    pub payment_method: Option<String>,
```
[the remark this answers is not in the document]
Comment (manoj-juspay) [reply crates/router/src/connector/stripe/transformers.rs:1040]: this is mandate_id it is as per our current norm

File: crates/router/src/core/payments/flows/authorize_flow.rs (hunk 9 above)
Comment (manoj-juspay) [reply crates/router/src/core/payments/flows/authorize_flow.rs:434]: i will add comment , its just a customer and call_customer check , have divided already in small fn

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/flows/authorize_flow.rs [hunk 13]
```diff
@@ -367,6 +378,168 @@ impl mandate::MandateBehaviour for types::PaymentsAuthorizeData {
...
+                    Some(connector_cust_id)
+                }
+                _ => None,
+            },
+            Err(err) => {
+                logger::debug!(payment_method_tokenization_error=?err);
+                None
+            }
+        };
+        update_connector_customer_in_customers(
```
[the remark this answers is not in the document]
Comment (manoj-juspay) [reply crates/router/src/core/payments/flows/authorize_flow.rs]: done

**Review by jagan-jaya — APPROVED** [review]

File: crates/router/src/core/payments/flows/authorize_flow.rs (hunk 9 above)
Comment (jarnura) [reply crates/router/src/core/payments/flows/authorize_flow.rs:418]: Add comments for None and else cases.

**Review by jarnura — CHANGES_REQUESTED** [review]
Add required labels.

**Review by jarnura — APPROVED** [review]

**Review by jarnura — DISMISSED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**